src/pages/common/_crsfdevice_page.c

```diff
@@ -0,0 +1,489 @@
+/*
+ This project is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ Deviation is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Deviation.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#if HAS_CRSF_CONFIG
+
+#include "crsf.h"
+
+#define CRSF_MAX_PARAMS  45   // one extra required, max observed is 40 in Nano RX
+#ifndef EMULATOR
+crsf_param_t crsf_params[CRSF_MAX_PARAMS];
+#else
+#include "protocol/crsf_params_tx._c"
+#include "protocol/crsf_params_rx._c"
+crsf_param_t *crsf_params = tx_crsf_params;
+#endif
+
+static struct crsfdevice_page * const mp = &pagemem.u.crsfdevice_page;
+static struct crsfdevice_obj * const gui = &gui_objs.u.crsfdevice;
+
+static u32 last_update;
+static u32 read_timeout;
+static u16 current_selected = 0;
+static u8 current_folder = 0;
+static u8 params_loaded;     // if not zero, number displayed so far for current device
+static u8 device_idx;   // current device index
+static u8 next_param;   // parameter and chunk currently being read
+static u8 next_chunk;
+
+static struct {
+    crsf_param_t *param;
+    u32 time;
+    u16 timeout;
+} command;
+
+#define CRSF_MAX_CHUNK_SIZE   58   // 64 - header - type - destination - origin
+#define CRSF_MAX_CHUNKS        4   // not in specification. Max observed is 3 for Nano RX
+static char recv_param_buffer[CRSF_MAX_CHUNKS * CRSF_MAX_CHUNK_SIZE];
+static char *recv_param_ptr;
+
+#define SEND_MSG_BUF_SIZE  16
+static u8 send_msg_buffer[SEND_MSG_BUF_SIZE];
+static volatile int send_msg_buf_count;     // tx data available semaphore with CRSF protocol
+
+#define CRSF_MAX_STRING_BYTES  2200     // max observed is 2010 in Nano RX
+#define CRSF_STRING_BYTES_AVAIL(current)  (CRSF_MAX_STRING_BYTES-((char *)(current)-strings))
+static char strings[CRSF_MAX_STRING_BYTES];
+static char *next_string = strings;
```

Comment:
This is a huge string. Can we move this to pagemem to save memory?

Reply:
Moving to pagemem saves 570 bytes.  I'll push the change after testing.

Reply:
Maybe other variables which logically only belong to the crsf page could move to pagemem as well. But since the ram save is "only" 570 bytes we wouldn't gain anything, because pagemem union is already dictated by char strings[CRSF_MAX_STRING_BYTES]

Reply:
Yes, it's just a lot of string space needed for all the parameters.  I've emailed TBS to see if they'll share information on the maximum space needed by each device.

For the transmitters where it's likely to be used there's still lots of space left :)